src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala

```diff
@@ -543,3 +450,277 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
+    }
+
+    // we are only interested in the class references in the descriptor, so we can skip over
+    // primitives and the brackets of array descriptors
+    def visitDescriptor(desc: String): Unit = (desc.charAt(0): @switch) match {
+      case '(' =>
+        val internalNames = mutable.ListBuffer.empty[String]
+        var i = 1
+        while (i < desc.length) {
+          if (desc.charAt(i) == 'L') {
+            val start = i + 1 // skip the L
+            while (desc.charAt(i) != ';') i += 1
+            internalNames += desc.substring(start, i)
+          }
+          // skips over '[', ')', primitives
+          i += 1
+        }
+        internalNames foreach visitInternalName
+
+      case 'L' =>
+        visitInternalName(desc.substring(1, desc.length - 1))
+
+      case '[' =>
+        visitInternalNameOrArrayReference(desc)
+
+      case _ => // skip over primitive types
+    }
+
+    def visitConstant(const: AnyRef): Unit = const match {
+      case t: Type => visitDescriptor(t.getDescriptor)
+      case _ =>
+    }
+
+    // in principle we could references to annotation types, as they only end up as strings in the
+    // constant pool, not as class references. however, the java compiler still includes nested
+    // annotation classes in the innerClass table, so we do the same. explained in detail in the
+    // large comment in class BTypes.
+    def visitAnnotation(annot: AnnotationNode): Unit = {
+      visitDescriptor(annot.desc)
+      if (annot.values != null) annot.values.asScala foreach visitConstant
+    }
+
+    def visitAnnotations(annots: java.util.List[_ <: AnnotationNode]) = if (annots != null) annots.asScala foreach visitAnnotation
+    def visitAnnotationss(annotss: Array[java.util.List[AnnotationNode]]) = if (annotss != null) annotss foreach visitAnnotations
+
+    def visitHandle(handle: Handle): Unit = {
+      visitInternalNameOrArrayReference(handle.getOwner)
+      visitDescriptor(handle.getDesc)
+    }
+  }
+
+  abstract class GenericSignatureVisitor {
+    def visitInternalName(internalName: InternalName): Unit
+
+    def raiseError(msg: String, sig: String, e: Option[Throwable] = None): Unit
+
+    def visitClassSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.classSignature() }
+    }
+
+    def visitMethodSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.methodSignature() }
+    }
+
+    def visitFieldSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.fieldSignature() }
+    }
+
+    private final class Parser(sig: String) {
+      private var index = 0
+      private val end = sig.length
+
+      private val Aborted: Throwable = new NoStackTrace { }
+      private def abort(): Nothing = throw Aborted
+
+      def safely(f: => Unit): Unit = try f catch {
```

Comment:
I assume this is intended to be inlined, perhaps add the annotation to express the desire more strongly.